File: DemoApp/DemoApp/CreateAdUnitIdViewController.m
//
//  CreateAdUnitIdViewController.m
//  DemoApp
//
//  Created by Glispa GmbH on 16.06.17.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

#import "CreateAdUnitIdViewController.h"
#import "Constants.h"

@interface CreateAdUnitIdViewController (DataSource) <UITableViewDelegate, UITableViewDataSource, UIPickerViewDelegate, UIPickerViewDataSource>
- (void)saveAdUnitsToUserDefaults;
- (void)readAdUnitsFromUserDefaults;
@end


@interface CreateAdUnitIdViewController () <UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet UITableView *adUnitsTableView;
@property (weak, nonatomic) IBOutlet UIPickerView *adTypesPicker;
@property (weak, nonatomic) IBOutlet UITextField *adUnitIdTextField;
@property (weak, nonatomic) IBOutlet UIImageView *backgroundImageView;

@property (strong, nonatomic) NSMutableArray <AdUnitId*> *adUnits;
@property (strong, nonatomic) NSArray <NSNumber*> *adTypes;

@end


@implementation CreateAdUnitIdViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupUI];
    [self setupData];
}

#pragma mark - Actions

- (IBAction)closeClicked:(UIButton *)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}


- (IBAction)createClicked:(UIButton *)sender {
    [self createNewAdUnitId];
}

#pragma mark - <UITextFieldDelegate>

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    return YES;
}

#pragma mark - Private

- (void)setupUI {
    self.backgroundImageView.image = _backgroundImage;
}

- (void)setupData {
    self.adTypes = @[@(AdBanner), @(AdInterstitial), @(AdVideo), @(AdNative),  @(AdNativeAssets)];
    [self readAdUnitsFromUserDefaults];
}

- (BOOL)isValidAdUnitId:(NSString*)adUnitId {
    NSError *error = nil;
    NSString *allowedSymbols  = @"^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$";
    NSRegularExpression* regex = [NSRegularExpression regularExpressionWithPattern: allowedSymbols options:0 error:&error];
    return [regex firstMatchInString:adUnitId options:0 range: NSMakeRange(0, [adUnitId length])] != nil;
}

- (void)createNewAdUnitId {
    if ([self isValidAdUnitId:self.adUnitIdTextField.text]) {
        NSNumber *adType = self.adTypes[[self.adTypesPicker selectedRowInComponent:0]];
        AdUnitId *adUnit = [AdUnitId unitIdWith:self.adUnitIdTextField.text withAdType:(AdType)[adType integerValue]];
        
        NSUInteger index = [self.adUnits indexOfObjectPassingTest:^BOOL(AdUnitId * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            return [obj.adUnitId isEqualToString:adUnit.adUnitId];
        }];
        
        if (index == NSNotFound) {
            [self.adUnits addObject:adUnit];
            [self saveAdUnitsToUserDefaults];
        }
        
        if (self.backBlock) {
            self.backBlock(adUnit.adUnitId, adUnit.adType);
        }
        
        [self dismissViewControllerAnimated:YES completion:nil];
    } else {
        UIAlertController *alertVC = [UIAlertController alertControllerWithTitle:@"Invalid format of adUnitId" message:@"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil];
        [alertVC addAction:okAction];
        [self presentViewController:alertVC animated:YES completion:nil];
    }
}

@end


@implementation CreateAdUnitIdViewController (DataSource)

- (void)saveAdUnitsToUserDefaults {
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject: self.adUnits];
    [[NSUserDefaults standardUserDefaults] setObject:data forKey:@"savedAdUnitsArray"];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

- (void)readAdUnitsFromUserDefaults {
    NSData *data = (NSData*)[[NSUserDefaults standardUserDefaults] objectForKey:@"savedAdUnitsArray"];
    NSArray *array = (NSArray*)[NSKeyedUnarchiver unarchiveObjectWithData:data];
    
    if (!array) {
        array = [self defaulAdUnitIds];
    }
    
    self.adUnits = [NSMutableArray arrayWithArray:array];
}


- (NSArray <AdUnitId*> *)defaulAdUnitIds {
    return @[[AdUnitId unitIdWith:DefaultAdUnitId.bannerAdUnitID withAdType:AdBanner],
             [AdUnitId unitIdWith:DefaultAdUnitId.interstitialAdUnitID withAdType:AdInterstitial],
             [AdUnitId unitIdWith:DefaultAdUnitId.videoAdUnitID withAdType:AdVideo],
             [AdUnitId unitIdWith:DefaultAdUnitId.nativeAdUnitID withAdType:AdNative],
             [AdUnitId unitIdWith:DefaultAdUnitId.nativeAssetsAdUnitID withAdType:AdNativeAssets]
            ];
}

#pragma mark - TableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.adUnits.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    
    AdUnitId *item = self.adUnits[indexPath.row];
    cell.textLabel.text = item.adUnitId;
    cell.detailTextLabel.text = [self adTypeStringValue:item.adType];
    
    cell.backgroundColor = [UIColor clearColor];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    AdUnitId *item = self.adUnits[indexPath.row];
    
    if (self.backBlock) {
        self.backBlock(item.adUnitId, item.adType);
    }
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - PickerView

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return self.adTypes.count;
}

- (NSString*)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return [self adTypeStringValue:(AdType)[self.adTypes[row] integerValue]];
}

- (NSString*)adTypeStringValue:(AdType)adType {
    switch (adType) {
        case AdBanner:
            return @"banner";
            
        case AdInterstitial:
            return @"interstitial";
            
        case AdVideo:
            return @"video";
            
        case AdNative:
            return @"native ad";
            
        case AdNativeAssets:
            return @"native assets ad";
            
        default:
            return @"unknown";
    }
}
@end


@implementation AdUnitId

+ (instancetype)unitIdWith:(NSString*)adUnitId withAdType:(AdType)adType {
    AdUnitId *instance = [self new];
    instance.adUnitId = adUnitId;
    instance.adType = adType;
    return instance;
}

- (void)encodeWithCoder:(NSCoder *)aCoder {
    [aCoder encodeObject:self.adUnitId forKey:@"adUnitId"];
    [aCoder encodeObject:@(self.adType) forKey:@"adUnitType"];
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super init]) {
        self.adUnitId = [aDecoder decodeObjectForKey:@"adUnitId"];
        self.adType = (AdType)[[aDecoder decodeObjectForKey:@"adUnitType"] integerValue];
    }
    return self;
}
@end

File: DemoApp/DemoAppSwift/CreateAdUnitIdViewController.swift
//
//  CreateAdUnitIdViewController.swift
//  DemoApp
//
//  Created by Glispa GmbH on 08.06.17.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

import UIKit

enum AdType: String {
    case banner
    case interstitial
    case video
    case native
    case nativeassets
    case unknown
}

class AdUnitId: NSObject, NSCoding {
    var adUnitId: String?
    var adType: AdType = .unknown
    
    convenience init(adUnitId: String, adType type: AdType) {
        self.init()
        self.adUnitId = adUnitId
        self.adType = type
    }

    func encode(with aCoder: NSCoder) {
        aCoder.encode(adUnitId, forKey: "adUnitId")
        aCoder.encode(adType.rawValue, forKey: "adUnitType")
    }

    required convenience init(coder decoder: NSCoder) {
        self.init()
        adUnitId = decoder.decodeObject(forKey: "adUnitId") as? String
        adType = AdType(rawValue: decoder.decodeObject(forKey: "adUnitType") as! String)!
    }
}

class CreateAdUnitIdViewController: UIViewController {
    @IBOutlet weak var adUnitsTableView: UITableView!
    @IBOutlet weak var adTypesPicker: UIPickerView!
    @IBOutlet weak var backgroundImageView: UIImageView!
    @IBOutlet weak var adUnitIdTextField: UITextField!

    var backgroundImage: UIImage?

    var backBlock: ((String, AdType) -> Void)?

    fileprivate var adUnits: [AdUnitId]?
    fileprivate let adTypes: [AdType] = [.banner, .interstitial, .video, .native, .nativeassets]

    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        setupData()
    }

    private func setupUI() {
        fillBackgroundImageView()
    }

    private func setupData() {
        readAdUnitsFromUserDefaults()
    }

    private func fillBackgroundImageView() {
        guard let image = backgroundImage else {
            return
        }
        backgroundImageView.image = image
    }

// MARK: - Action

    @IBAction func closeClicked(_ sender: UIButton) {
        self.dismiss(animated: true)
    }

    @IBAction func createClicked(_ sender: UIButton) {
        createNewAdUnitId()
    }
    
// MARK: - Privatre
    
    private func createNewAdUnitId() {
        if let text = adUnitIdTextField.text, isValidAdUnitId(text) {
            let adUnit = AdUnitId(adUnitId: text, adType: adTypes[adTypesPicker.selectedRow(inComponent: 0)])
    
            if adUnits?.filter({$0.adUnitId == text}).count == 0 {
                adUnits?.append(adUnit)
                saveAdUnitsToUserDefaults()
            }
    
            backBlock?(adUnit.adUnitId!, adUnit.adType)
            self.dismiss(animated: true)
        } else {
            let alert = UIAlertController(title: "Invalid format of adUnitId", message: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx", preferredStyle: .alert)
            let okAction = UIAlertAction(title: "Ok", style: .cancel, handler: nil)
            alert.addAction(okAction)
            self.present(alert, animated: true, completion: nil)
        }
    }
}

extension CreateAdUnitIdViewController {
    func isValidAdUnitId(_ adUnitId: String?) -> Bool {
        let allowedSymbols = "^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$"
        let allowedRegEx = NSPredicate(format:"SELF MATCHES %@", allowedSymbols)
        guard allowedRegEx.evaluate(with: adUnitId) else { return false }
        return true
    }
}

extension CreateAdUnitIdViewController: UITableViewDelegate, UITableViewDataSource {

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return adUnits?.count ?? 0
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)

        guard let item = adUnits?[indexPath.row] else { return cell }
        cell.textLabel?.text = item.adUnitId
        cell.detailTextLabel?.text = item.adType.rawValue

        cell.backgroundColor = UIColor.clear
        return cell
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let item = adUnits?[indexPath.row] else { return }
        backBlock?(item.adUnitId!, item.adType)
        self.dismiss(animated: true)
    }
}

extension CreateAdUnitIdViewController {
    private func adUnitsKey() -> String {
        return "savedAdUnitsArray"
    }

    fileprivate func saveAdUnitsToUserDefaults() {
        let data = NSKeyedArchiver.archivedData(withRootObject: adUnits!)
        UserDefaults.standard.set(data, forKey: adUnitsKey())
    }

    fileprivate func readAdUnitsFromUserDefaults() {

        guard let data = UserDefaults.standard.object(forKey: adUnitsKey()) as? Data else {
            adUnits = defaulAdUnitIds()
            return
        }

        guard let array = NSKeyedUnarchiver.unarchiveObject(with: data) as? [AdUnitId] else {
            adUnits = defaulAdUnitIds()
            return
        }

        adUnits = array
    }
    
    fileprivate func defaulAdUnitIds() -> [AdUnitId] {
        return [AdUnitId(adUnitId: DefaultAdUnitId.bannerAdUnitID, adType: .banner),
                AdUnitId(adUnitId: DefaultAdUnitId.interstitialAdUnitID, adType: .interstitial),
                AdUnitId(adUnitId: DefaultAdUnitId.videoAdUnitID, adType: .video),
                AdUnitId(adUnitId: DefaultAdUnitId.nativeAdUnitID, adType: .native),
                AdUnitId(adUnitId: DefaultAdUnitId.nativeAssetsAdUnitID, adType: .nativeassets)]
    }
}

extension CreateAdUnitIdViewController: UIPickerViewDelegate, UIPickerViewDataSource {

    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }

    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return adTypes.count
    }

    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return adTypes[row].rawValue
    }
}

extension CreateAdUnitIdViewController: UITextFieldDelegate {
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
}
